### ISSUE TITLE
Two results from an overnight run: the model is data-limited, and the hybrid traces only pay off with frequent state resets

### ISSUE THREAD

**Issue by SimonVitzethum** [jrz97619761/test-model-thing#23]
I ran two 4 h arms on one laptop GPU last night, identical except for
`traces`. The traces question was the one I set out to answer; the more useful
result is the one that fell out of comparing against the earlier 8 h run.

**Summary**

1. Training *less* on the same 95 MB gave a **better** held-out model. 14
   epochs reach 1.63 BPB held out, 24 epochs reached 1.82. At this size the
   model is data-limited, not capacity-limited, which changes what is worth
   trying next.
2. Carrying exact traces across window boundaries buys nothing over a long
   stream at equal wall-clock time, but is **7.3 % better** when the recurrent
   state is reset every 128 bytes. The trade is legible, and it depends on how
   the model will be used.

---

## 1. The model is data-limited

Same architecture (`dim=512 layers=16 experts=8 topk=2`), same data, same
laptop GPU:

| run | bytes seen | epochs over 95 MB | training BPB | **held-out BPB** | gap |
|---|---|---|---|---|---|
| earlier 8 h run | 2.31 GB | 24.3 | 1.241 | 1.824 | 0.58 |
| last night, 4 h | 1.34 GB | 14.1 | 1.331 | **1.637** | 0.31 |

Half the compute, and the held-out number improves by 0.19 BPB. The training
loss goes the other way, and the gap between the two nearly halves. That is
overfitting on a 95 MB corpus, plainly.

The practical consequence: more parameters would mostly widen that gap. What
helps here is more signal per byte or more data — multi-token prediction, a
better optimizer, a larger corpus — not a bigger model. I would also read the
earlier "8 h run reached 2.50 BPB" number in the README as saying more about
the epoch count than about the architecture.

If anyone has been scaling this model up and seeing less than they expected,
this is worth checking first: how many times has your model seen the corpus?

---

## 2. The hybrid traces: an A/B at equal wall-clock time

The open question from the earlier discussion was whether carrying exact
traces across TBPTT window boundaries (`traces=1`) is worth its cost. The
gradient cosines said the traces recover most of what truncation loses,
especially for the gate. What they cannot say is whether that becomes a better
model per unit of compute, because the traces also cost throughput.

Two arms, same architecture, same seed, same data, **same wall-clock time**,
the only difference being `traces`. Equal time rather than equal steps is the
honest comparison: the cheaper arm may convert its saved time into more
updates. Each arm got its own `decaysteps`, matched to the steps it was
expected to reach, so both saw a complete cosine schedule.

```
dim=512 layers=16 experts=8 topk=2 batch=16 seqlen=128
aux=0.1 half_max=65536 warmup=2000
arm A: traces=1  decaysteps=562000
arm B: traces=0  decaysteps=620000
data: enwik8, 95 MB train / 5 MB held out, RTX 5070 Laptop (8 GB), 4 h per arm
```

| | `traces=1` | `traces=0` |
|---|---|---|
| throughput | 93 123 B/s | **94 958 B/s** (+2.0 %) |
| optimizer steps in 4 h | 654 630 | 667 648 |
| bytes seen | 1.341 GB | 1.367 GB |
| training BPB (run average) | 1.3314 | 1.3348 |
| **held-out BPB, full context** | 1.6367 | **1.6292** |
| held-out, state reset every 128 B | **2.1907** | 2.3643 |
| held-out, state reset every 512 B | **1.8144** | 1.8579 |
| held-out, state reset every 4096 B | **1.6492** | 1.6590 |

**At full context the traces do not pay for themselves.** The difference is
0.5 % of BPB in favour of plain TBPTT, which is about what the throughput
advantage buys in extra steps. With one seed per arm I would not claim the
ordering is real; what I would claim is that the traces buy no visible
advantage over a long stream.

**With frequent state resets they clearly do.** Resetting every 128 bytes
costs the trace-trained model 0.554 BPB and the TBPTT-trained one 0.735 BPB,
a 7.3 % gap; at 512 bytes it is still 2.3 %. That is consistent with what the
traces are: gradient signal for what the state should carry *into* a window. A
model trained with that signal leans less on a long warm-up of its own state.

So: for a long-lived stream, window-truncated gradients are enough and the
2 % is better spent on steps. For a model that starts cold often — short
documents, chat turns, per-request state — the traces earn their cost.

### The gradient cosines do not predict the outcome

Measured with `gradcheck` on each arm's own checkpoint, against full BPTT over
4 x 4096 bytes, window 128:

| | on the `traces=1` model | on the `traces=0` model |
|---|---|---|
| decay, truncated / hybrid | 0.848 / 0.864 | 0.941 / 0.959 |
| gate, truncated / hybrid | 0.836 / **0.973** | 0.891 / **0.978** |
| embedding, truncated / hybrid | 0.829 / 0.880 | 0.807 / 0.898 |

The hybrid cosines are high on both checkpoints, and the *truncated* cosine is
markedly worse on the trace-trained model (0.848 vs 0.941 for the decay).
Training with traces moves the weights into a region where windowed gradients
approximate full BPTT less well. A cosine of 0.97 against full BPTT is
apparently not the quantity that decides the outcome — which is worth
remembering before optimizing that number further.

---

## Reproducing

```sh
timeout -s INT 14400s ./train enwik8/train.bin hybrid.ckpt \
    dim=512 layers=16 experts=8 topk=2 batch=16 seqlen=128 aux=0.1 \
    traces=1 half_max=65536 warmup=2000 decaysteps=562000 steps=0 saveevery=5000
./train enwik8/held.bin hybrid.ckpt mode=eval
./train enwik8/held.bin hybrid.ckpt mode=eval maxcarry=128
./gradcheck enwik8/held.bin hybrid.ckpt len=4096 window=128 seqs=4
```

Caveat on reproducibility: the gradients end in atomic adds, so two runs of the
same command do not give bit-identical checkpoints. These are single runs per
arm; treat differences below about 1 % as noise. Making the gradients
order-independent is on my list, exactly so that comparisons like this one can
be repeated instead of trusted.

**Comment by jrz97619761** [comment]
can you run a train with the full English Wikipedia dump ([on dumps.wikimedia.org, there are multiple files, use all of them](https://dumps.wikimedia.org/other/mediawiki_content_current/enwiki/2026-09-01/xml/bzip2/)) *in addition to* the Simple English Wikipedia dump? if you want, add in the other english wikimedia sister sites (wikibooks, wikinews, wikiquote, wikisource, wikiversity, wikivoyage, wikitionary) that are in the dumps as well.

compare the results to the simple english dump and see if we are indeed data limited. thanks!
